src/proto/streams/prioritize.rs

```diff
@@ -266,6 +266,27 @@ impl Prioritize {
         Ok(())
     }
 
+    /// Reclaim all capacity assigned to the stream and re-assign it to the
+    /// connection
+    pub fn reclaim_all_capacity(&mut self, stream: &mut store::Ptr) {
+        let available = stream.send_flow.available().as_size();
+        stream.send_flow.claim_capacity(available);
+        // Re-assign all capacity to the connection
+        self.assign_connection_capacity(available, stream);
+    }
+
+    /// Reclaim just reserved capacity, not buffered capacity, and re-assign
+    /// it to the connection
+    pub fn reclaim_reserved_capacity(&mut self, stream: &mut store::Ptr) {
+        // only reclaim requested capacity that isn't already buffered
+        if stream.requested_send_capacity > stream.buffered_send_data {
```

Comment:
IIRC, `requested_send_capacity` is capacity that has been **requested** but not actually allocated to the stream, i.e., there was not enough connection capacity to provide.

I would think that you would need to check `stream.send_flow.available()` just like above, but then subtracting `buffered_send_data` from that?

If I am correct, we'd probably want to add a test to catch this case.

Reply:
I don't think that is correct, it looks like here that `requested >= buffered`: https://github.com/carllerche/h2/blob/master/src/proto/streams/prioritize.rs#L146